base/inference.jl

```diff
@@ -1479,7 +1529,20 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
                 end
             end
             pc´ = pc+1
-            if isa(stmt,GotoNode)
+            if isa(changes,StateUpdate) && isa((changes::StateUpdate).var, GenSym)
+                changes = changes::StateUpdate
+                id = (changes.var::GenSym).id+1
+                new = changes.vtype
+                old = gensym_types[id]
+                if old===NF || !(new <: old)
+                    gensym_types[id] = tmerge(old, new)
+                    for r in gensym_uses[id]
+                        if !is(s[r],()) # s[r] === () => unreached statement
```

Comment:
This change is really unnecessarily complicated. It adds the `gensym_uses` data structure, plus a pass to populate it. This is redundant, since this whole algorithm is already about finding uses of everything, in a flow-sensitive way. gensym_uses finds uses in a flow-*in*sensitive way, which only works because of assumptions about how the gensyms are used.

In short I think it's simpler to handle GenSyms just how symbols are handled now: store a type at each program point, and then afterwards in `type_annotate` move the information to the var info arrays.